pandas/core/internals/blocks.py

```diff
@@ -2066,7 +2067,7 @@ def _can_hold_element(self, element: Any) -> bool:
                 and not issubclass(tipo.type, (np.datetime64, np.timedelta64))
                 and self.dtype.itemsize >= tipo.itemsize
             )
-        return is_integer(element)
+        return is_integer(element) or (is_float(element) and element.is_integer())
```

Comment:
@jbrockmendel are you sure you think the patch should be *here*; this is hit by a lot of paths where this doesn't make any sense. we almost always want to take some sort of action based on return value here if it fails. e.g. why is this not more specific to replace?

Reply:
Also is this simply wrong:
```python
[ins] In [3]: is_integer(1.0)
Out[3]: False
```
especially since
```python
[ins] In [4]: 1.0.is_integer()
Out[4]: True
```

Reply:
ahh its doing an equality check under the hood, we are not

```
n [60]: 1.0.is_integer()                                                                                                                                  
Out[60]: True

In [61]: 1.1.is_integer()                                                                                                                                  
Out[61]: False
```

Reply:
maybe could replace this entirely with `element.is_integer()` and call it a day

Reply:
though my concern is *still* that the purpose of this check is *can* we hold an integer, not are we a float and *can* be treated as an integer. though practically it may not make a difference.

Reply:
So I don't realy understand ...
Need I replace `return is_integer(element)` into `return element.is_integer()`.
If I do that, need I try except the `AttributeError` and return `None` ?

Reply:
Or `False` ?

Reply:
Longer-term we're going to have to look more carefully about can_hold_element and what we want it to mean.  In the short run, for this check to be consistent with the tipo check above it should exclude floats

(so my previous suggestion may have been unhelpful, sorry)

Reply:
ok this is fine @QuentinN42 if you can add a comment here would be good